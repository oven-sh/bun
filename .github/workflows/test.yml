name: Test

concurrency:
  group: ${{ github.workflow }}-${{ inputs.run-id || github.run_id }}
  cancel-in-progress: true

on:
  push:
  workflow_call:
    inputs:
      run-id:
        type: string
        required: true

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/download-artifact@v4
        with:
          run-id: ${{ inputs.run-id || '8623134412' }}
          pattern: bun-*
      - run: |
          ls -la
