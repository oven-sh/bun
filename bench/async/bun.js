import { bench, run } from "mitata";

bench("sync", () => {});
bench("async", async () => {});
bench("await 1", async () => await 1);

await run();
