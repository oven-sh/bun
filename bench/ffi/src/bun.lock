{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@napi-rs/cli": "^2.10.1",
        "@node-rs/helper": "^1.3.3",
      },
    },
  },
  "packages": {
    "@napi-rs/cli": ["@napi-rs/cli@2.14.8", "", { "bin": { "napi": "scripts/index.js" } }, "sha512-IvA3s8BqohMdUbOkFn7+23u1dhIJZCkA8Xps7DD4SLdCMbcbUF6MUuKiqxuqmVHBFTaxU25sU56WdX3efqGgPw=="],

    "@napi-rs/triples": ["@napi-rs/triples@1.1.0", "", {}, "sha512-XQr74QaLeMiqhStEhLn1im9EOMnkypp7MZOwQhGzqp2Weu5eQJbpPxWxixxlYRKWPOmJjsk6qYfYH9kq43yc2w=="],

    "@node-rs/helper": ["@node-rs/helper@1.3.3", "", { "dependencies": { "@napi-rs/triples": "^1.1.0" } }, "sha512-p4OdfQObGN9YFy5WZaGwlPYICQSe7xZYyXB0sxREmvj1HzGKp5bPg2PlfgfMZEfnjIA882B9ZrnagYzZihIwjA=="],
  }
}
