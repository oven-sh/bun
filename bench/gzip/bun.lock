{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@babel/standalone": "7.24.10",
      },
    },
  },
  "packages": {
    "@babel/standalone": ["@babel/standalone@7.24.10", "", {}, "sha512-nGC37EKfmelpyCXto1pw6SBkD5ZQRdMbL6WISi28xWit9dtiy9dChU1WgEfzturUTxrmOGkMDRrCydFMA7uOaQ=="],
  }
}
