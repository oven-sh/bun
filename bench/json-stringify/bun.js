import { bench, run } from "mitata";

bench("JSON.stringify({hello: 'world'})", () =>
  JSON.stringify({ hello: "world" }),
);

const otherUint8Array = new Uint8Array(1024);
bench("Uint8Array.from(otherUint8Array)", () =>
  Uint8Array.from(otherUint8Array),
);

run();
