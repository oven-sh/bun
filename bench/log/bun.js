import { bench, run } from "mitata";

bench("console.log('hello')", () => console.log("hello"));
bench("console.log({ hello: 'object' })", () => console.log({ hello: "object" }));
await run();
