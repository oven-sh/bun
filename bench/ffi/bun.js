import { ptr, dlopen, CString, toBuffer } from "bun:ffi";
import { run, bench, group } from "mitata";

const { napiNoop, napiHash, napiString } = require(import.meta.dir + "/src/ffi_napi_bench.node");

const {
  symbols: {
    ffi_noop: { native: ffi_noop },
    ffi_hash: { native: ffi_hash },
    ffi_string: { native: ffi_string },
  },
} = dlopen(import.meta.dir + "/src/ffi_napi_bench.node", {
  ffi_noop: { args: [], returns: "void" },
  ffi_string: { args: [], returns: "ptr" },
  ffi_hash: { args: ["ptr", "u32"], returns: "u32" },
});

const bytes = new Uint8Array(64);

group("bun:ffi", () => {
  bench("noop", () => ffi_noop());
  bench("hash", () => ffi_hash(ptr(bytes), bytes.byteLength));

  bench("c string", () => new CString(ffi_string()));
});

if (process.env.SHOW_NAPI)
  group("bun:napi", () => {
    bench("noop", () => napiNoop());
    bench("hash", () => napiHash(bytes));

    bench("string", () => napiString());
  });

await run();
