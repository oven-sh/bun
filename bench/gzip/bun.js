import { run, bench, group } from "mitata";
import { gzipSync, gunzipSync } from "bun";

const data = new TextEncoder().encode("Hello World!".repeat(9999));

const compressed = gzipSync(data);

const libraries = ["zlib"];
if (Bun.semver.satisfies(Bun.version, ">=1.1.21")) {
  libraries.push("libdeflate");
}

const options = { library: undefined };
const benchFn = (name, fn) => {
  if (libraries.length > 1) {
    group(name, () => {
      for (const library of libraries) {
        bench(library, () => {
          options.library = library;
          fn();
        });
      }
    });
  } else {
    bench(name, () => {
      fn();
    });
  }
};

benchFn(`roundtrip - "Hello World!".repeat(9999))`, () => {
  gunzipSync(gzipSync(data, options), options);
});

benchFn(`gzipSync("Hello World!".repeat(9999)))`, () => {
  gzipSync(data, options);
});

benchFn(`gunzipSync("Hello World!".repeat(9999)))`, () => {
  gunzipSync(compressed, options);
});

await run();
