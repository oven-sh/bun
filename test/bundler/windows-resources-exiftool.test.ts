import { spawn } from "bun";
import { describe, expect, test } from "bun:test";
import { bunEnv, bunExe, isWindows, tempDirWithFiles } from "harness";
import { join } from "path";

// Skip these tests on Windows as they're for verifying cross-compilation
describe.skipIf(isWindows)("Windows Resource Editing with exiftool", () => {
  const hasExiftool = Bun.which("exiftool") !== null;

  // Common build function
  async function buildWindowsExecutable(
    dir: string,
    outfile: string,
    windowsOptions: Record<string, string | boolean> = {},
  ) {
    const args = [
      bunExe(),
      "build",
      "--compile",
      "--target=bun-windows-x64-v1.2.19",
      ...Object.entries(windowsOptions).flatMap(([key, value]) =>
        value === true ? [`--${key}`] : [`--${key}`, value as string],
      ),
      join(dir, "index.js"),
      "--outfile",
      join(dir, outfile),
    ];

    await using proc = spawn({
      cmd: args,
      cwd: dir,
      env: bunEnv,
      stderr: "pipe",
    });

    const [stderr, exitCode] = await Promise.all([new Response(proc.stderr).text(), proc.exited]);
    if (exitCode !== 0) {
      console.error("Build failed with exit code:", exitCode);
      console.error("stderr:", stderr);
    }
    expect(exitCode).toBe(0);

    // Filter out mimalloc warnings which are expected for large allocations (Windows PE files ~118MB)
    const filteredStderr = stderr
      .split("\n")
      .filter(
        line =>
          !line.includes("mimalloc: warning:") &&
          !line.includes("(this may still be a valid very large allocation") &&
          !line.includes("(yes, the previous pointer") &&
          line.trim() !== "",
      )
      .join("\n")
      .trim();
    expect(filteredStderr).toBe("");

    return join(dir, outfile);
  }

  // Test icon data (minimal valid ICO file)
  const createTestIcon = () => {
    // ICO header (6 bytes)
    const header = Buffer.from([
      0x00,
      0x00, // Reserved
      0x01,
      0x00, // Type (1 = ICO)
      0x01,
      0x00, // Count (1 icon)
    ]);

    // Directory entry (16 bytes)
    const dirEntry = Buffer.from([
      0x10, // Width (16)
      0x10, // Height (16)
      0x00, // Color count (0 = 256 colors)
      0x00, // Reserved
      0x01,
      0x00, // Planes
      0x08,
      0x00, // Bit count
      0x28,
      0x01,
      0x00,
      0x00, // Bytes in resource (296)
      0x16,
      0x00,
      0x00,
      0x00, // Image offset (22)
    ]);

    // Minimal BMP data (just a header for simplicity)
    const bmpHeader = Buffer.alloc(40);
    bmpHeader.writeUInt32LE(40, 0); // Header size
    bmpHeader.writeInt32LE(16, 4); // Width
    bmpHeader.writeInt32LE(32, 8); // Height (double for AND mask)
    bmpHeader.writeUInt16LE(1, 12); // Planes
    bmpHeader.writeUInt16LE(8, 14); // Bit count
    bmpHeader.writeUInt32LE(0, 16); // Compression
    bmpHeader.writeUInt32LE(256, 20); // Image size

    // Create minimal image data
    const imageData = Buffer.alloc(256); // 16x16x8bpp

    return Buffer.concat([header, dirEntry, bmpHeader, imageData]);
  };

  test.skipIf(!hasExiftool)("verifies version info with exiftool", async () => {
    const dir = tempDirWithFiles("exiftool-test", {
      "index.js": `console.log("Testing with exiftool");`,
    });

    const exePath = await buildWindowsExecutable(dir, "test.exe", {
      "windows-version": "9.8.7.6",
      "windows-description": "My Custom Description",
      "windows-publisher": "Test Publisher Inc",
      "windows-title": "My Custom Product",
      "windows-copyright": "Copyright 2024 Test Publisher Inc",
    });

    try {
      // Run exiftool to extract metadata
      await using proc = spawn({
        cmd: ["exiftool", "-j", exePath],
        stdout: "pipe",
        stderr: "pipe",
      });

      const [stdout, stderr, exitCode] = await Promise.all([
        new Response(proc.stdout).text(),
        new Response(proc.stderr).text(),
        proc.exited,
      ]);

      expect(exitCode).toBe(0);
      expect(stderr).toBe("");

      console.log("Raw exiftool output:", stdout);
      console.log("Executable path for debugging:", join(dir, "test.exe"));
      const metadata = JSON.parse(stdout)[0];

      // Verify the version information
      expect(metadata.FileVersionNumber).toBe("9.8.7.6");
      expect(metadata.ProductVersionNumber).toBe("9.8.7.6");
      expect(metadata.FileDescription).toBe("My Custom Description");
      expect(metadata.CompanyName).toBe("Test Publisher Inc");
      expect(metadata.ProductName).toBe("My Custom Product");
      expect(metadata.LegalCopyright).toBe("Copyright 2024 Test Publisher Inc");
    } finally {
      await Bun.file(exePath).unlink();
    }
  });

  test.skipIf(!hasExiftool)("verifies subsystem change with exiftool", async () => {
    const dir = tempDirWithFiles("exiftool-subsystem", {
      "index.js": `console.log("Testing subsystem");`,
    });

    const exePath = await buildWindowsExecutable(dir, "hidden.exe", {
      "windows-hide-console": true,
    });

    try {
      await using proc = spawn({
        cmd: ["exiftool", "-Subsystem", exePath],
        stdout: "pipe",
      });

      const [stdout, exitCode] = await Promise.all([new Response(proc.stdout).text(), proc.exited]);
      expect(exitCode).toBe(0);

      // Windows GUI subsystem
      expect(stdout).toContain("Windows GUI");
    } finally {
      await Bun.file(exePath).unlink();
    }
  });

  test.skipIf(!hasExiftool)("verifies icon resource with exiftool", async () => {
    const dir = tempDirWithFiles("exiftool-icon", {
      "index.js": `console.log("Testing icon");`,
      "icon.ico": createTestIcon(),
    });

    const exePath = await buildWindowsExecutable(dir, "icon.exe", {
      "windows-icon": join(dir, "icon.ico"),
      "windows-version": "1.0.0.0",
    });

    try {
      await using proc = spawn({
        cmd: ["exiftool", "-j", exePath],
        stdout: "pipe",
      });

      const [stdout, exitCode] = await Promise.all([new Response(proc.stdout).text(), proc.exited]);
      expect(exitCode).toBe(0);

      const metadata = JSON.parse(stdout)[0];

      // Even with an icon, the version should still be set
      expect(metadata.FileVersionNumber).toBe("1.0.0.0");
      expect(metadata.ProductVersionNumber).toBe("1.0.0.0");
    } finally {
      await Bun.file(exePath).unlink();
    }
  });

  test.skipIf(!hasExiftool)("verifies all fields with exiftool", async () => {
    const dir = tempDirWithFiles("exiftool-all", {
      "index.js": `console.log("Testing all fields");`,
      "icon.ico": createTestIcon(),
    });

    const exePath = await buildWindowsExecutable(dir, "all.exe", {
      "windows-icon": join(dir, "icon.ico"),
      "windows-version": "5.4.3.2",
      "windows-description": "Complete Test Application",
      "windows-publisher": "Acme Corporation",
      "windows-title": "Acme Test Suite",
      "windows-copyright": "(c) 2024 Acme Corporation. All rights reserved.",
      "windows-hide-console": true,
    });

    try {
      await using proc = spawn({
        cmd: ["exiftool", "-j", exePath],
        stdout: "pipe",
      });

      const [stdout, exitCode] = await Promise.all([new Response(proc.stdout).text(), proc.exited]);
      expect(exitCode).toBe(0);

      const metadata = JSON.parse(stdout)[0];

      // Verify all fields
      expect(metadata.FileVersionNumber).toBe("5.4.3.2");
      expect(metadata.ProductVersionNumber).toBe("5.4.3.2");
      expect(metadata.FileDescription).toBe("Complete Test Application");
      expect(metadata.CompanyName).toBe("Acme Corporation");
      expect(metadata.ProductName).toBe("Acme Test Suite");
      expect(metadata.LegalCopyright).toBe("(c) 2024 Acme Corporation. All rights reserved.");
      expect(metadata.Subsystem).toBe("Windows GUI");
    } finally {
      await Bun.file(exePath).unlink();
    }
  });

  test.skipIf(!hasExiftool)("snapshot test with exiftool", async () => {
    const dir = tempDirWithFiles("exiftool-snapshot", {
      "index.js": `console.log("Snapshot test");`,
    });

    const exePath = await buildWindowsExecutable(dir, "snapshot.exe", {
      "windows-version": "1.2.3.4",
      "windows-description": "Snapshot Test App",
      "windows-publisher": "Snapshot Publisher",
      "windows-title": "Snapshot Product",
      "windows-copyright": "Copyright 2024",
    });

    try {
      // Get full exiftool output for snapshot
      await using proc = spawn({
        cmd: ["exiftool", exePath],
        stdout: "pipe",
      });

      const [stdout, exitCode] = await Promise.all([new Response(proc.stdout).text(), proc.exited]);
      expect(exitCode).toBe(0);

      // Extract relevant version info fields for snapshot
      const versionFields = stdout
        .split("\n")
        .filter(
          line =>
            line.includes("File Version Number") ||
            line.includes("Product Version Number") ||
            line.includes("File Description") ||
            line.includes("Company Name") ||
            line.includes("Product Name") ||
            line.includes("Legal Copyright") ||
            line.includes("File Version") ||
            line.includes("Product Version"),
        )
        .join("\n");

      expect(versionFields).toMatchSnapshot();
    } finally {
      await Bun.file(exePath).unlink();
    }
  });

  test.skipIf(!hasExiftool)("verifies real ICO file with exiftool", async () => {
    // Real ICO file created with ImageMagick (full multi-size version)
    const realIcoPath = join(import.meta.dir, "real-icon.ico");
    if (!(await Bun.file(realIcoPath).exists())) {
      // Skip if real icon file doesn't exist
      return;
    }

    const dir = tempDirWithFiles("exiftool-real-ico", {
      "index.js": `console.log("Testing real ICO");`,
      "real.ico": await Bun.file(realIcoPath).bytes(),
    });

    const exePath = await buildWindowsExecutable(dir, "realico.exe", {
      "windows-icon": join(dir, "real.ico"),
      "windows-version": "2.0.0.0",
      "windows-title": "Real Icon Test",
    });

    try {
      await using proc = spawn({
        cmd: ["exiftool", "-j", exePath],
        stdout: "pipe",
      });

      const [stdout, exitCode] = await Promise.all([new Response(proc.stdout).text(), proc.exited]);
      expect(exitCode).toBe(0);

      const metadata = JSON.parse(stdout)[0];

      // Verify version is still set with real icon
      expect(metadata.FileVersionNumber).toBe("2.0.0.0");
      expect(metadata.ProductName).toBe("Real Icon Test");
    } finally {
      await Bun.file(exePath).unlink();
    }
  });
});
