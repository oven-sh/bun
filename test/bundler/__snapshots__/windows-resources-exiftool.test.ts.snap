// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Windows Resource Editing with exiftool snapshot test with exiftool 1`] = `
"File Version Number             : 1.2.3.4
Product Version Number          : 1.2.3.4
Company Name                    : Snapshot Publisher
File Description                : Snapshot Test App
File Version                    : 1.2.3.4
Legal Copyright                 : Copyright 2024
Product Name                    : Snapshot Product
Product Version                 : 1.2.3.4"
`;
